### PR TITLE
Introducing SalesforceLogger - a new logging framework for MobileSDK components

### PR DESCRIPTION
**Pull request by bhariharan**
This pull request adds a new logging framework in `MobileSDK`.

Features in this PR:
- Allows logging to console and to file.
- Log levels for console and file.
- Logging at the component level.
- Allows different log levels for each component.
- Allows a customizable circular buffer size of log file.
- One log file per component.
- File logger uses `Square's [Tape](http://square.github.io/tape/)` framework for fast I/O.

Coming soon:

- Dynamic tuning of log levels per component through custom attributes from connected apps.
- Retrieval of the log files for upload.
- Persistence of logging preferences in `SharedPrefs`.

### CODE REVIEW

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java [hunk 1]
```diff
@@ -0,0 +1,257 @@
...
+    private int maxSize;
+
+    /**
+     * Parameterized constructor.
+     *
+     * @param context Context.
+     * @param componentName Component name.
+     * @throws IOException If file initialization was unsuccessful.
+     */
+    public FileLogger(Context context, String componentName) throws IOException {
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java]: One logger per component.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+    public int getMaxSize() {
+        return maxSize;
+    }
+
+    /**
+     * Sets the maximum number of log lines that can be stored in this file.
+     *
+     * @param size Maximum number of log lines.
+     */
+    public synchronized void setMaxSize(int size) {
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java]: Allows the app to specify max size of circular buffer.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java [hunk 3]
```diff
@@ -0,0 +1,293 @@
...
+    private static final String TAG = "SalesforceLogger";
+    private static final String LOG_LINE_FORMAT = "LEVEL: %s, TAG: %s, MESSAGE: %s";
+    private static final String LOG_LINE_FORMAT_WITH_EXCEPTION = "LEVEL: %s, TAG: %s, MESSAGE: %s, EXCEPTION: %s";
+    private static final ExecutorService THREAD_POOL = Executors.newFixedThreadPool(3);
+    private static Map<String, SalesforceLogger> LOGGERS;
+
+    /**
+     * An enumeration of log levels.
+     */
+    public enum Level {
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java]: Log levels are pretty much same as stock `Android` log levels.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java [hunk 4]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    /**
+     * Wipes all components currently being logged. Used only by tests.
+     */
+    public synchronized static void flushComponents() {
+        LOGGERS = null;
+    }
+
+    private SalesforceLogger(String componentName, Context context) {
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java]: One logger per component.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java [hunk 5]
```diff
@@ -0,0 +1,293 @@
...
+        this.context = context;
+        if (isDebugMode()) {
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java]: If the `APK` is a debug binary, the default level is `DEBUG`. If it's a release binary, the default level is `ERROR`.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java [hunk 6]
```diff
@@ -0,0 +1,293 @@
...
+    }
+
+    /**
+     * Logs the message passed in at the level specified.
+     *
+     * @param level Log level.
+     * @param tag Log tag.
+     * @param message Log message.
+     */
+    public void log(Level level, String tag, String message) {
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java]: Simple `log` API. Logs to console using stock `Android` logger and then writes the log line to file as well.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java [hunk 7]
```diff
@@ -0,0 +1,293 @@
...
+
+    /**
+     * Logs the message and throwable passed in at the level specified.
+     *
+     * @param level Log level.
+     * @param tag Log tag.
+     * @param message Log message.
+     * @param e Throwable to be logged.
+     */
+    public void log(Level level, String tag, String message, Throwable e) {
```
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java]: Overloaded API with a `Throwable`.

File: libs/test/SalesforceAnalyticsTest/src/com/salesforce/androidsdk/analytics/logger/FileLoggerTest.java [hunk 8]
```diff
@@ -0,0 +1,318 @@
...
+
+import java.util.ArrayList;
+import java.util.List;
+
+/**
+ * Tests for FileLogger.
+ *
+ * @author bhariharan
+ */
+public class FileLoggerTest extends InstrumentationTestCase {
```
Comment (bhariharan) [inline libs/test/SalesforceAnalyticsTest/src/com/salesforce/androidsdk/analytics/logger/FileLoggerTest.java]: Tests for file logging.

File: libs/test/SalesforceAnalyticsTest/src/com/salesforce/androidsdk/analytics/logger/SalesforceLoggerTest.java [hunk 9]
```diff
@@ -0,0 +1,104 @@
...
+import android.test.InstrumentationTestCase;
+
+import java.util.Set;
+
+/**
+ * Tests for SalesforceLogger.
+ *
+ * @author bhariharan
+ */
+public class SalesforceLoggerTest extends InstrumentationTestCase {
```
Comment (bhariharan) [inline libs/test/SalesforceAnalyticsTest/src/com/salesforce/androidsdk/analytics/logger/SalesforceLoggerTest.java]: Tests for the wrapper logger framework.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java [hunk 10]
```diff
@@ -0,0 +1,293 @@
...
+            default:
+                Log.d(tag, message, e);
+        }
+        if (level != Level.OFF) {
+            logToFile(level, tag, message, e);
+        }
+    }
+
+    private void logToFile(final Level level, final String tag, final String message, final Throwable e) {
+        THREAD_POOL.execute(new Runnable() {
```
Comment (wmathurin) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java]: Could log lines in the file end up being in a different order from the ones in the console?

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java [hunk 11]
```diff
@@ -0,0 +1,257 @@
...
+ *
+ * @author bhariharan
+ */
+public class FileLogger {
+
+    private static final String LOG_SUFFIX = "_log";
+    private static final String UTF8 = "UTF-8";
+    private static final String ASCII = "US-ASCII";
+    private static final String TAG = "FileLogger";
+    private static final int MAX_SIZE = 1000;
```
Comment (wmathurin) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java]: Should the default be bigger?

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java [hunk 12]
```diff
@@ -0,0 +1,257 @@
...
+        return logLine;
+    }
+
+    /**
+     * Returns a list of log lines in FIFO and removes them as they are being read.
+     *
+     * @param numLines Number of lines to be read.
+     * @return List of log lines.
+     */
+    public List<String> readAndRemoveLogLinesAsList(int numLines) {
```
Comment (wmathurin) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java]: Why do we need that?

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java (hunk 11 above)
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java]: Sure - I'll make it `10000` for now.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java (hunk 12 above)
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/FileLogger.java]: The idea is to provide a way to retrieve log lines for apps that want to upload them.

File: libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java (hunk 10 above)
Comment (bhariharan) [inline libs/SalesforceAnalytics/src/com/salesforce/androidsdk/analytics/logger/SalesforceLogger.java]: I don't think so. The underlying file write API in `Tape` is `synchronized`, so order is guaranteed.

**Comment by bhariharan** [comment]
I'll add the remaining features in the next PR, so that it's easy to review.

**Pull request merged**